### ISSUE TITLE
Review Windows String code

### ISSUE THREAD

**Issue by DanRStevens** [OutpostUniverse/op2ext#148]
Windows native raw strings are used in a few places in the code base. A few have been updated recently. We should perhaps take the time to look over the rest while it's still fresh. In particular, I'd like to update `ConvertLPWToString` from StringConversion.cpp, which hasn't been touched yet.

----

Quick summary:

Windows native raw strings are used in the following files:
srcStatic/StringConversion.h
srcStatic/StringConversion.cpp
srcStatic/GetCommandLineArguments.cpp
srcStatic/WindowsErrorCode.cpp
srcDLL/DllMain.cpp

Windows native raw strings form part of the public function signature for:
`ConvertLPWToString` - Converts Windows native raw string from wide to narrow, and then wraps in `std::string`
`ConvertLpctstrToString` - Wraps a Windows native raw string into `std::string`/`std::wstring` (changes/additions pending)

Windows native raw strings are processed internally by:
`GetCommandLineArguments` - Parsed command line arguments
`GetLastErrorString` - Win32 API error string
`LoadLibraryNew` - DLL name in event hook to tell when OP2Shell.dll is loaded

----

The `ConvertLPWToString` uses a `const LPWSTR pw` parameter. Windows provides the type `LPCWSTR` which expands to `const LPWSTR`, so we're being a bit inconsistent here.

----

The `<windows.h>` header is a big file, which can slow compilation, so we should try to minimize it's use when possible, particularly in header files. Using Windows specific types requires an include for `<windows.h>`. In particular, functions that use these types as part of their signature need to include `<windows.h>` in the header file for that function.

To reduce header pollution, functions that need Windows specific types in their signatures can be split into different files from other methods that don't use Windows specific types as part of their signature. This should reduce the transitive impact of header file includes.

Specifically, `ConvertLPWToString` and `ConvertLpctstrToString` can be moved to a separate file.

----

The `ConvertLPWToString` uses raw `new[]`/`delete[]` calls. We should switch that to either use `unique_ptr`, or perhaps a `std::string` or `std::vector` as the dynamic buffer. Using a `std::string` may even reduce the amount of copying needed to return a `std::string`. The would make the code cleaner, more robust to editing, and exception safe.

----

The signature of `ConvertLPWToString` returns a `std::string` through an out parameter and reports errors through a `bool` return value. We should perhaps use the more idiomatic C++ way, where we return the `std::string`, and report errors with exceptions.

----

The `ConvertLPWToString` method has effectively two responsibilities. One is to convert from wide strings to narrow strings. The other is to wrap the narrow string into an `std::string`. We already have the `ConvertLpctstrToString` method to wrap Windows raw strings into `std::string` (or `std::wstring`, change pending). Perhaps we should untie the two responsibilities, and re-use the other method.

In particular, a raw wide string could first be wrapped into an `std::wstring`, and then converted to a narrow `std::string`. The wrapped string buffers could potentially be passed directly to the [`WideCharToMultiByte`](https://docs.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-widechartomultibyte) Windows API method.

----

There is an alternative way to convert between wide and narrow strings. Some C++ standard library methods can help here. Some relevant info can be found at:
[How to convert wstring into string?](https://stackoverflow.com/questions/4804298/how-to-convert-wstring-into-string)
One answer gives step by step code to convert wide strings to narrow, and then gives a one-liner for converting the opposite direction from narrow to wide:
```
std::wstring str = std::wstring_convert<std::codecvt_utf8<wchar_t>>().from_bytes("some string");
```

In the example they make use of the standard library method [`std::wstring_convert`](https://en.cppreference.com/w/cpp/locale/wstring_convert) from the `<locale>` header. Unfortunately this method has been deprecated in C++17, though as one note points out, it's been moved to an annex, where it will remain until a suitable replacement is built. As such it is still effectively part of the standard, and will remain so for now.

Related to the internals of the encoding conversions is the [`<codecvt>`](https://en.cppreference.com/w/cpp/header/codecvt) header file, which is part of the Localization library. It provides a few standard conversions between the utf8/utf16 encodings.

**Comment by DanRStevens** [comment]
Oh wow, I just learned something fascinating about another dark corner of C++ relating to `typedef`s.

Apparently `const LPWSTR` doesn't do what you expect.
```
using T = wchar_t*;

EXPECT_FALSE((std::is_same<const T, const wchar_t*>::value));
EXPECT_FALSE((std::is_same<const T, wchar_t const*>::value));
EXPECT_TRUE((std::is_same<const T, wchar_t* const>::value));

EXPECT_FALSE((std::is_same<T const, const wchar_t*>::value));
EXPECT_FALSE((std::is_same<T const, wchar_t const*>::value));
EXPECT_TRUE((std::is_same<T const, wchar_t* const>::value));
```

It seems a `const LPWSTR` is actually the same as a `LPWSTR const`. That is, the pointer is const, not the `wchar_t`. You have to use `LPCWSTR` to get a pointer to a `const wchar_t`.

When I first tried to write some unit tests, I was getting a rather unexpected warning from the compiler. A quick Google search lead me to this:
[LPCWSTR vs. const LPWSTR](https://groups.google.com/forum/#!topic/microsoft.public.vc.language/v468FXdwJKI)

Just wow.

So yeah, this declaration:
```
bool ConvertLPWToString(std::string& stringOut, const LPWSTR pw, UINT codepage = CP_ACP);

```
With test code:
```
std::string output;
EXPECT_TRUE(ConvertLPWToString(output, L"Hello world"));
EXPECT_EQ("Hello world", output);
```
Was producing the warning:

> warning: ISO C++ forbids converting a string constant to ‘LPWSTR {aka wchar_t*}’ [-Wwrite-strings]
>   EXPECT_TRUE(ConvertLPWToString(output, L"Hello world"));
>

**Comment by DanRStevens** [comment]
Most things discussed here have now been merged. There is still a couple of outstanding items.

I did not yet look into using the standard C++ library to do the string conversions between wide and narrow strings. I am fairly happy with the current Windows API code that handles this, so I'm considering just skipping this entirely.

The other is the use of `<windows.h>` from the StringConversion.h header file. This is needed to handle the signatures of the methods:
```
std::string ConvertWideToNarrow(std::wstring_view inputWideString, UINT codepage = CP_ACP);

std::string ConvertLpctstrToString(LPCSTR str);
std::wstring ConvertLpctstrToString(LPCWSTR str);
```

In particular, the symbols `UINT`, `CP_ACP`,`LPCSTR`, and `LPCWSTR` are used, which all come from `<windows.h>`. Perhaps some changes could remove those dependencies, or as suggested earlier, the methods could move to their own header file.
